planner/core/task.go

```diff
@@ -1812,6 +1899,348 @@ func (p *PhysicalHashAgg) attach2TaskForMpp1Phase(mpp *mppTask) task {
+	// customized proj here rather than depend on postOptimize to insert a blunt one for us.
+	//
+	// proj4Partial output all the base col from lower op + caseWhen proj cols.
+	proj4Partial := new(PhysicalProjection).Init(p.ctx, mpp.p.statsInfo(), mpp.p.SelectBlockOffset())
+	for _, col := range mpp.p.Schema().Columns {
+		proj4Partial.Exprs = append(proj4Partial.Exprs, col)
+	}
+	proj4Partial.SetSchema(mpp.p.Schema().Clone())
+
+	partialHashAgg := partialAgg.(*PhysicalHashAgg)
+	partialHashAgg.GroupByItems = append(partialHashAgg.GroupByItems, groupingIDCol)
+	partialHashAgg.schema.Append(groupingIDCol.(*expression.Column))
+	// it will create a new stats for partial agg.
+	partialHashAgg.scaleStats4GroupingSets(groupingSets, groupingIDCol.(*expression.Column), proj4Partial.Schema(), proj4Partial.statsInfo())
+	for _, fun := range partialHashAgg.AggFuncs {
+		if !fun.HasDistinct {
+			// for normal agg phase1, we should also modify them to target for specified group data.
+			eqExpr := expression.NewFunctionInternal(p.ctx, ast.EQ, types.NewFieldType(mysql.TypeTiny), groupingIDCol, expression.NewUInt64Const(fun.GroupingID))
+			caseWhen := expression.NewFunctionInternal(p.ctx, ast.Case, fun.Args[0].GetType(), eqExpr, fun.Args[0], expression.NewNull())
+			caseWhenProjCol := &expression.Column{
+				UniqueID: p.ctx.GetSessionVars().AllocPlanColumnID(),
+				RetType:  fun.Args[0].GetType(),
+			}
+			proj4Partial.Exprs = append(proj4Partial.Exprs, caseWhen)
+			proj4Partial.Schema().Append(caseWhenProjCol)
+			fun.Args[0] = caseWhenProjCol
+		}
+	}
+
+	// step2: adjust middle agg
+	// proj4Middle output all the base col from lower op + caseWhen proj cols.（reuse partial agg stats）
+	proj4Middle := new(PhysicalProjection).Init(p.ctx, partialHashAgg.stats, partialHashAgg.SelectBlockOffset())
+	// middleHashAgg shared the same stats with the final agg does.
+	middleHashAgg := cloneHashAgg
+	middleSchema := expression.NewSchema()
+	schemaMap := make(map[int64]*expression.Column, len(middleHashAgg.AggFuncs))
+	for _, col := range partialHashAgg.Schema().Columns {
+		proj4Middle.Exprs = append(proj4Middle.Exprs, col)
+	}
+	proj4Middle.SetSchema(partialHashAgg.Schema().Clone())
+	for _, fun := range middleHashAgg.AggFuncs {
+		col := &expression.Column{
+			UniqueID: p.ctx.GetSessionVars().AllocPlanColumnID(),
+			RetType:  fun.RetTp,
+		}
+		if fun.HasDistinct {
+			fun.Mode = aggregation.Partial1Mode
+			// change the middle distinct agg args to target for specified group data.
+			// even for distinct agg multi args like count(distinct a,b), either null of a,b will cause count func to skip this row.
+			// so we only need to set the either one arg of them to be a case when expression to target for its self-group:
+			// Expr = (case when groupingID = ? then arg else null end)
+			// count(distinct a,b) => count(distinct (case when groupingID = targetID then a else null end), b)
```

Comment:
Hmm, I remembered we don't need the `case when` here, at least for built-in agg functions.

Reply:
it's a standard way, actually, we can remove this projection to let distinct-agg and normal-agg aggregate on whole scope data. (null rows included but not counted for these null-strict built aggs), what do you think?

Reply:
Sounds good. count(*) is an exception.

Reply:
count(*) has already been rewritten as count(1) when entering here